### PR TITLE
chore: analyze language identification models performance on short ingredient texts with precision-recall evaluation (#349)

### PR DESCRIPTION
**Pull request by korablique**
This is the research conducted in the next issue: https://github.com/openfoodfacts/openfoodfacts-ai/issues/349

`01_extract_data.py`: extracts all texts with their languages from [huggingface dataset](https://huggingface.co/datasets/openfoodfacts/product-database).

`02_select_short_texts_with_known_ingredients.py`: filters texts with length up to 10 words, performs ingredient analysis by OFF API, selects ingredient texts with at least 80% of known ingredients, adds short texts from manually checked data.

What is manually checked data: \
I created a validation dataset from texts from OFF (42 languages, 15-30 texts per language).
I took 30 random texts in each language, obtained language predictions using the Deepl API and two other models ([language-detection-fine-tuned-on-xlm-roberta-base](https://huggingface.co/ivanlau/language-detection-fine-tuned-on-xlm-roberta-base) and [multilingual-e5-language-detection](https://huggingface.co/Mike0307/multilingual-e5-language-detection)). For languages they don’t support, I used Google Translate and ChatGPT for verification. (As a result, after correcting the labels, some languages have fewer than 30 texts).

`03_calculate_metrics.py`: obtains predictions by [FastText](https://huggingface.co/facebook/fasttext-language-identification) and [lingua language detector](https://github.com/pemistahl/lingua-py) models for texts up to 10 words long, and calculates precision, recall and f1-score.

Results are in files: 10_words_metrics.csv, fasttext_confusion_matrix.csv, lingua_confusion_matrix.csv.

It turned out that both models demonstrate low precision and high recall for some languages (indicating that the threshold might be too high and should be adjusted).

### CODE REVIEW

**Review by baslia — COMMENTED** [review]
I think that would be great to have a separate file (like 04_inference.py) dedicated for inference, then it would be easy to warp the code and deploy it in the future

**Comment by korablique** [comment]
@baslia Could you please help me with this error? https://github.com/openfoodfacts/openfoodfacts-ai/actions/runs/12106320879/job/33751933052?pr=365
I haven't worked with labeler before, don't understand how to fix this issue

**Comment by baslia** [comment]
> @baslia Could you please help me with this error? https://github.com/openfoodfacts/openfoodfacts-ai/actions/runs/12106320879/job/33751933052?pr=365 I haven't worked with labeler before, don't understand how to fix this issue

Hey, I thought this was just a check about the PR label attribute, I just attached the label "enhancement".
But this isn't working, but I don't think this test is a big deal

**Comment by raphael0202** [comment]
Yes indeed it's a configuration issue from the repo, it's safe to ignore!
I haven't have time to review your work yet, I will find some time to do it next week!
Thank you for pushing the code.

File: language_identification/scripts/01_extract_data.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+from datasets import load_dataset
+
+
+def main():
+    # extracting all texts with their languages from huggingface dataset
+
+    # path where to save selected data
+    dataset_file = os.path.join(Path(__file__).parent, "texts_with_lang.csv")
+
+    hf_dataset = load_dataset("openfoodfacts/product-database", split=f"food")
```
Comment (raphael0202) [inline language_identification/scripts/01_extract_data.py]:
```suggestion
    hf_dataset = load_dataset("openfoodfacts/product-database", split="food")
```

File: language_identification/scripts/01_extract_data.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    data = set()
+    for entry, main_lang in zip(hf_dataset["ingredients_text"], hf_dataset["lang"]):  # iterate over products
+        for product_in_lang in entry:
+            if product_in_lang["text"]:
+                lang = main_lang if product_in_lang["lang"] == "main" else product_in_lang["lang"]
+                data.add((product_in_lang["text"], lang))
+
+    df = pd.DataFrame(data, columns=["ingredients_text", "lang"])
+    df.dropna(inplace=True)
+    df.to_csv(dataset_file, index=False)
```
Comment (raphael0202) [inline language_identification/scripts/01_extract_data.py:24]: I had an issue when generating the CSV:
`_csv.Error: need to escape, but no escapechar set`
By providing an escape char, it works:
`df.to_csv(dataset_file, index=False, escapechar="\\")`

**Comment by raphael0202** [comment]
Thank you a lot for this PR!
Lingua results look much better than fasttext, in almost all languages :)

I will add a few suggestions for the next steps in the main issue #349

**Review by raphael0202 — APPROVED** [review]

**Pull request merged**